nifi-commons/nifi-site-to-site-client/src/main/java/org/apache/nifi/remote/client/http/HttpClient.java

```diff
@@ -168,6 +169,9 @@ public Transaction createTransaction(final TransferDirection direction) throws H
             try {
                 transactionUrl = apiClient.initiateTransaction(direction, portId);
                 commSession.setUserDn(apiClient.getTrustedPeerDn());
+            } catch (final NoContentException e) {
+                logger.debug("Peer {} has no flowfiles to provide", peer);
+                throw e;
```

Comment:
If the remote NiFi is a cluster and there are multiple NiFi nodes, even if this client get NoContentException from node1, node2 may have queued data to transfer. I think we should keep iterating next peer in this case, instead of throwing the exception.

If no node has data, this method eventually returns `null` transaction. From callers, it looks the same with the case where all nodes are penalized. I think that is acceptable.
If we take this approach, we don't have to catch `NoContentException` at callers such as StandardRemoteGroupPort.

Reply:
I partly agree. Checking all the nodes is a fair point, although transaction == null is considered as an error case:
```
        if (transaction == null) {
            logger.debug("{} Unable to create transaction to communicate with; all peers must be penalized, so yielding context", this);
            session.rollback();
            context.yield();
            return;
        }
```
Neither the comment here, nor the rollback should apply in case peers has no data to provide. We should simply yield.

Reply:
I totally agree with you on the necessity of separating the two cases. How about changing SiteToSiteClient.createTransaction semantics as follows?
- AS IS
    - HandshakeException, PortNotRunningException, UnknownPortException in irregular case
    - Returns a Transaction to use for sending or receiving data, or `null` if all nodes are penalized.
- TO BE (by this PR)
    - HandshakeException, PortNotRunningException, UnknownPortException in irregular case
    - Or (new) NoPeerAvailableException when all peers are penalized. Because this is an error case. We need to change [SocketClient](https://github.com/apache/nifi/blob/master/nifi-commons/nifi-site-to-site-client/src/main/java/org/apache/nifi/remote/client/socket/SocketClient.java#L129), too.
    - Returns a Transaction to use for sending or receiving data, or `null` if communication was successful but no transaction is needed, ex. no data to transfer.

This allows callers to handle each cases correctly, and the API looks more intuitive. How do you think?

Reply:
Generally I like the idea, the impact is going to be way more than the Jira itself would justify, but I'm fine with that :)